### ISSUE TITLE
Problem with actions unrender, render

### ISSUE THREAD

**Issue by esbhome** [ractivejs/ractive#2527]
**Description of the bug:**
Hi. I use my own component with derective {{>content}} and when I unrender and then render {{>content}} lost - as result I have empty.
Example - http://jsfiddle.net/juh826sa/6/
Initial view - http://prntscr.com/avz3bi
How we can see a bug? 
    1) hit unrender, 
     2) hit render (nothing has changed, but in console we see that structure of table was rendered, but with empty cells http://prntscr.com/avz0a0)
Please help me. How I can use {{>content}} or maybe something else I should use.
**Versions affected:**
Ractive.js 0.7.3
**Platforms affected:**

**Reproduction:**
http://jsfiddle.net/juh826sa/6/
**Code:**
&lt;script id="master" type="text/ractive"&gt;
    &lt;mytable&gt;
    	{{#if(k == 'first')}}
      	&lt;span style='color:red'&gt;{{.}}&lt;/span&gt;
       {{else}}
       	{{.}}
       {{/if}}
    &lt;/mytable&gt;
&lt;/script&gt;

&lt;script id="mytable" type="text/ractive"&gt;
    &lt;table&gt;
    {{#list}}
    	&lt;tr&gt;
      	{{#.:k}}
      	&lt;td&gt;{{&gt;content}}&lt;/td&gt;
        {{/}}
      &lt;/tr&gt;
    {{/}}
    &lt;/table&gt;
&lt;/script&gt;

&lt;div id="test"&gt;&lt;/div&gt;
Ractive.components["mytable"] = Ractive.extend({
    template: "#mytable"
});

var oRactive = new Ractive({
    data: {
        list: [....]
    },
    el: "#test",
    template: "#master"
});

**Comment by fskreuz** [comment]
A temporary workaround would be to use `ractive.detach` and `ractive.insert`. Not sure how `unrender/render` and `detach/insert` differ in this case tho.

**Comment by esbhome** [comment]
Thanks, with ractive.detach and ractive.insert works. http://jsfiddle.net/juh826sa/8/

**Comment by esbhome** [comment]
How I can view my instanse detached or inseted?

**Comment by esbhome** [comment]
Found. oRactive.detached. When detach() in this var place all from DOM when insert - this var is null.
,

**Comment by esbhome** [comment]
With oRactive.detach() I have problem too - data lost when I insert and instace falling off DOM and when I do monipulation render else instance (work) and if I do detach()  and then insert() then all work fine.

**Comment by esbhome** [comment]
Hmm in fiddle I cant repeat my problem :( http://jsfiddle.net/juh826sa/13/

**Comment by esbhome** [comment]
But on my syte event onrender dont fire when I do insert()

**Comment by esbhome** [comment]
O! this is repeat problem http://jsfiddle.net/juh826sa/14/

**Comment by esbhome** [comment]
This interest the else variant of bug http://jsfiddle.net/juh826sa/15/.
hit second, first, change, change

**Comment by esbhome** [comment]
If I user another partul like td_content instead content ({{>td_content}}) I have the same result. http://jsfiddle.net/juh826sa/16/

**Comment by esbhome** [comment]
Problem may be in this, http://prntscr.com/awo2ws after rerendering all components concated to one element. On this place I see this http://prntscr.com/awo41z

**Comment by evs-chris** [comment]
I'm a little bit lost, but would the `detach` and `insert` hooks (events) help? It seems that you were trying to use render and unrender and their corresponding hooks, but render and unrender in this scenario are apparently broken in 0.7. detach and insert work but don't trigger the render and unrender hooks because they have their own - `oninsert` and `ondetach`.

Alternatively, instead of manually rendering and unrendering instances, is there any way you can have the controller do it by putting them in conditional sections?

**Comment by esbhome** [comment]
I use "putting them in conditional sections", but and I use render and unrender. With detach and insert I have problem too but more then with {{>content}}. Now I simply removed code where I use {{>content}} and used just if condition but this ugli in my situation because I made my table as component and I want reuse it but small code (td rendering) always different and I used {{>content}} now I set var that switch different td view.
But I hope that will be fixed

**Comment by esbhome** [comment]
With detach I have problem with data after insert. Its insert without it or I dont know what heppens besause Ractive doesn't fail but part that I inserted not work but if I inserted something else and again that part looks fine but when I try switch view to another section - creating new DOM element and old present too.

**Comment by esbhome** [comment]
I try again repeat bug with detach and insert in fiddle.

**Comment by esbhome** [comment]
"detach and insert work but don't trigger the render and unrender hooks because they have their own - oninsert and ondetach."
Yes, but I did not update the data after insertion (onupdate event), I want to with the previous data has been inserted in all.

**Comment by esbhome** [comment]
I understand that I should use detach, but what the diferense beatween detach, unrender, teardown?

**Comment by esbhome** [comment]
And first of all thankyou for the answer.

**Comment by fskreuz** [comment]
It would be best to outline the actual use case on the original post rather than trying to solve the problem your current approach is giving you. There may be a bug in your current approach, yes, but at the same time people might suggest other, *better* alternatives to what you are trying to do... before this thread becomes unnecessarily long.

Related read: [What is the XY Problem?](http://meta.stackexchange.com/a/66378/177538)

**Comment by evs-chris** [comment]
All of those methods are really meant to be used on top-level instances, though they can be used with components with sometimes strange side-effects, as you've discovered. The main instance typically manages its own components. There has been some discussion on extending more API control over components in the near future though.

That said, here are the scenarios for those methods (as I see them :smile:):

`detach` - I want to move this instance somewhere else with `insert`
`unrender` - I'm done with this for now, but I want to use it again later with `render` - there is a bug here in latest that has been fixed in edge
`teardown` - I'm done with this forever

For the duplication, I did see that behavior, but the fiddle that was doing so was using render and unrender. When I changed it to detach and insert, the duplication stopped.

I'd recommend something more like:

```html
This is the main ractive template
<div id="here-are-the-components">
  {{#if !section}}
  <myTable list="{{someList}}">
    cell content 1
  </myTable>
  {{elseif section === 1}}
  <myTable list="{{someOtherList}}">
    different cell content 2
  </myTable>
</div>
<div id="here-are-the-controls">
  <button on-click="set('section', 0)">section 1</button>
  <button on-click="set('section', 1)">section 2</button>
</div>
```

That way the main ractive instance is managing the component lifecycle. Even better, you can just have a conditional for `style="display: none;"` that gets toggled with the section so that the component isn't completely removed. That seems to be closer to the scenario that you're shooting for in the fiddle.

**Comment by esbhome** [comment]
Ok. I have an app with 9 Rctive instance info block, menu, submenu, content (5 time), windows.
On some view (content 1) i need render a table, in view 2 (content 2) i have form and else... In content 3 i need table like in content 1 but cell little bit different, and I create a component. I switch content with render unrender and I found this problem.

**Comment by esbhome** [comment]
Thankyou for answer. I think like you. Detach when move component, unreder when I want render for content place another ractive (and user maybe return to the first). I do like this. About saction - I have different ractive .extand and differend file with template for different component. I have 9 ractive.extand() (and 9 folders with its files (css, model.js, component.js, tamplates.html)) for my components and their parts. And I cannt write it like one section. If I understand you right...

**Comment by evs-chris** [comment]
Ractive supports a fair amount of dynamism, so you can load components on the fly and use them with templates generated on the fly. You could use `resetPartial` or partial expressions along with whatever component loader you're using to render the component and then detach it and re-insert it elsewhere on the page if you need to do so. That would probably be slightly less error-prone than the process you're using now.

**Comment by esbhome** [comment]
Same problem -  http://jsfiddle.net/9g3pB/90/ but with decorators.

**Comment by esbhome** [comment]
And else problem with render unrender action http://jsfiddle.net/9g3pB/91/
If I wont use {{{ '<span> not escaped paste </span>' }}}, when I unrender and render - I have error. At that example you can hit button second and then button first - and you catch error... if in the template remove {{{ }}} and set {{ }}, error removed.

**Comment by evs-chris** [comment]
@esbhome I think #2595 may be helpful for you.

**Comment by esbhome** [comment]
I found the solution for fix "If I want use {{{ ' not escaped paste ' }}}, when I unrender and render - I have error."
http://jsfiddle.net/9g3pB/95/
if I before unrender hide part of template with {{{ '...' }}} and show after render (oncomlete) - I doesnt have an error :)

**Comment by esbhome** [comment]
I found beautiful solution for partuals - if I use local partuals - when render/unrender actions - I have an error (I wrote up), but if I use global Ractive.partials - I have no a problem with action render/unrender.
